File: Assets/Script/Example/SQLite.lua
-- [yue]: Script/Example/SQLite.yue
local DB = Dora.DB -- 1
local print = _G.print -- 1
local p = _G.p -- 1
local thread = Dora.thread -- 1
local pairs = _G.pairs -- 1
local threadLoop = Dora.threadLoop -- 1
local App = Dora.App -- 1
local ImGui = Dora.ImGui -- 1
local Vec2 = Dora.Vec2 -- 1
local result = DB:transaction({ -- 4
	"DROP TABLE IF EXISTS test", -- 4
	"CREATE TABLE test (id INTEGER PRIMARY KEY, value TEXT)", -- 5
	{ -- 7
		"INSERT INTO test VALUES(?, ?)", -- 7
		{ -- 9
			{ -- 9
				false, -- 9
				"hello" -- 9
			}, -- 9
			{ -- 10
				false, -- 10
				"world" -- 10
			}, -- 10
			{ -- 11
				false, -- 11
				"ok" -- 11
			} -- 11
		} -- 8
	} -- 6
}) -- 3
print(result and "Success" or "Failure") -- 16
print(DB:exist("test")) -- 18
p(DB:query("SELECT * FROM test", true)) -- 20
print("row changed:", DB:exec("DELETE FROM test WHERE id > 1")) -- 22
print("row changed:", DB:exec("UPDATE test SET value = ? WHERE id = 1", { -- 24
	"hello world!" -- 24
})) -- 24
local _anon_func_0 = function(items) -- 33
	local _accum_0 = { } -- 33
	local _len_0 = 1 -- 33
	for _index_0 = 1, #items do -- 33
		local item = items[_index_0] -- 33
		_accum_0[_len_0] = item[1] -- 33
		_len_0 = _len_0 + 1 -- 33
	end -- 33
	return _accum_0 -- 33
end -- 33
thread(function() -- 26
	print("insert async") -- 27
	local data -- 28
	do -- 28
		local _accum_0 = { } -- 28
		local _len_0 = 1 -- 28
		for k in pairs(_G) do -- 28
			_accum_0[_len_0] = { -- 28
				false, -- 28
				k -- 28
			} -- 28
			_len_0 = _len_0 + 1 -- 28
		end -- 28
		data = _accum_0 -- 28
	end -- 28
	p(DB:insertAsync("test", data)) -- 29
	print("query async...") -- 31
	local items = DB:queryAsync("SELECT value FROM test WHERE value NOT LIKE 'hello%' ORDER BY value ASC") -- 32
	return p(_anon_func_0(items)) -- 33
end) -- 26
print("OK") -- 35
local windowFlags = { -- 40
	"NoDecoration", -- 40
	"AlwaysAutoResize", -- 40
	"NoSavedSettings", -- 40
	"NoFocusOnAppearing", -- 40
	"NoNav", -- 40
	"NoMove" -- 40
} -- 40
return threadLoop(function() -- 48
	local width -- 49
	width = App.visualSize.width -- 49
	ImGui.SetNextWindowBgAlpha(0.35) -- 50
	ImGui.SetNextWindowPos(Vec2(width - 10, 10), "Always", Vec2(1, 0)) -- 51
	ImGui.SetNextWindowSize(Vec2(240, 0), "FirstUseEver") -- 52
	return ImGui.Begin("SQLite", windowFlags, function() -- 53
		ImGui.Text("SQLite (Yuescript)") -- 54
		ImGui.Separator() -- 55
		return ImGui.TextWrapped("Doing database operations in synchronous and asynchronous ways.") -- 56
	end) -- 56
end) -- 56

File: Tools/dora-rust/dora-test/src/tests/sqlite.rs
/* Copyright (c) 2016-2026 Li Jin <dragon-fly@qq.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE. */

use dora_ssr::*;

fn get_str(arr: &Array, index: i32) -> Option<String> {
	arr.get(index)?.into_str()
}

fn get_i32(arr: &Array, index: i32) -> Option<i32> {
	arr.get(index)?.into_i32()
}

pub fn test() {
	let mut sqls = DBQuery::new();
	sqls.add("DROP TABLE IF EXISTS test");
	sqls.add("CREATE TABLE test (id INTEGER PRIMARY KEY, value TEXT)");
	let mut params = DBParams::new();
	let mut arr = Array::new();
	let data = ["hello", "world", "ok"];
	for i in 0..data.len() {
		arr.add(false).add(data[i]);
		params.add(&arr);
		arr.clear();
	}
	sqls.add_with_params("INSERT INTO test VALUES(?, ?)", params);

	let result = DB::transaction(sqls);
	p!("transaction result: {}", result);
	p!("table test exist: {}", DB::exist("test"));

	let mut records = DB::query("SELECT * FROM test", true);
	records.read(&arr);
	for i in 0..arr.get_count() as i32 {
		if let Some(value) = get_str(&arr, i) {
			p!("{}", value);
		}
	}

	let mut count = 0;
	while records.read(&arr) {
		if let Some(value) = get_i32(&arr, 0) {
			p!("{}", value);
		}
		if let Some(value) = get_str(&arr, 1) {
			p!("{}", value);
		}
		count += 1;
	}
	p!("result count: {}", count);


	p!("row deleted: {}", DB::exec("DELETE FROM test WHERE id > 2"));
	let mut params = DBParams::new();
	params.add(&Array::new().add("hello world!"));
	p!("row updated: {}", DB::exec_with_records("UPDATE test SET value = ? WHERE id = 1", params));

	p!("insert async");
	let mut params = DBParams::new();
	let data = ["B", "A", "D", "C", "F", "E"];
	for i in 0..data.len() {
		arr.clear();
		arr.add(false).add(data[i]);
		params.add(&arr);
	}
	DB::insert_async("test", params, Box::new(|result| {
		p!("insert async result: {}", result);
		p!("query async...");
	}));

	DB::query_with_params_async("SELECT value FROM test WHERE value NOT LIKE 'hello%' ORDER BY value ASC", &Array::new(), true, Box::new(move |mut result| {
		while result.read(&arr) {
			if let Some(value) = get_str(&arr, 0) {
				p!("{}", value);
			}
		}
	}));

	p!("OK");

	let windows_flags =
		ImGuiWindowFlag::NO_DECORATION |
		ImGuiWindowFlag::AlwaysAutoResize |
		ImGuiWindowFlag::NoSavedSettings |
		ImGuiWindowFlag::NoFocusOnAppearing |
		ImGuiWindowFlag::NO_NAV |
		ImGuiWindowFlag::NoMove;
	let mut imgui_node = Node::new();
	imgui_node.schedule(Box::new(move |_| {
		let size = App::get_visual_size();
		ImGui::set_next_window_bg_alpha(0.35);
		ImGui::set_next_window_pos_opts(&Vec2::new(size.width - 10.0, 10.0), ImGuiCond::Always, &Vec2::new(1.0, 0.0));
		ImGui::set_next_window_size_opts(&Vec2::new(240.0, 0.0), ImGuiCond::FirstUseEver);
		ImGui::begin_opts("SQLite", windows_flags, || {
			ImGui::text("SQLite (Rust)");
			ImGui::separator();
			ImGui::text_wrapped("Doing database operations in synchronous and asynchronous ways.");
		});
		false
	}));
}